ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala

```diff
@@ -367,20 +389,38 @@ class ContractsService(
         private[this] def unsafeRunAsync[A](cio: doobie.ConnectionIO[A]) =
           dao.transact(cio).unsafeToFuture()
 
+        private[this] def timed[A](
+            timer: Timer,
+            it: doobie.ConnectionIO[A],
+        ): doobie.ConnectionIO[A] = {
+          for {
+            ctx <- doobie.free.connection.pure(timer.time())
+            res <- it
+            _ = ctx.stop()
```

Comment:
Side effects are dangerous.

```suggestion
            it: => doobie.ConnectionIO[A],
        ): doobie.ConnectionIO[A] = {
          for {
            _ <- fconn.pure(())
            ctx <- fconn.pure(timer.time())
            res <- it
            _ <- fconn.pure(ctx.stop())
```

Reply:
Good point! I assume that the argument is not lazy for pure then? Nevertheless in the end it shouldn't be seen as pure at all.

Reply:
Indeed, it is not pure.

Reply:
I'm honestly not 100% sure whether this is good enough or future-proof for the timer start even now. Flatmapped pures can be fused. Given the calling context, this is probably fine, though.

Putting something on the right of a flatmap rather than of a map, when the left side of the flatmap is not a pure, is always a good way to delay it though. So the timer stop is just fine with this change.